tmol/score/cartbonded/cartbonded_energy_term.py

```diff
@@ -0,0 +1,287 @@
+import torch
+import numpy
+
+from itertools import permutations
+
+from tmol.score.atom_type_dependent_term import AtomTypeDependentTerm
+
+from tmol.database import ParameterDatabase
+
+# from tmol.score.cartbonded.params import CartBondedGlobalParams
+from tmol.score.cartbonded.cartbonded_whole_pose_module import (
+    CartBondedWholePoseScoringModule,
+)
+
+from tmol.chemical.restypes import RefinedResidueType
+from tmol.pose.packed_block_types import PackedBlockTypes
+from tmol.pose.pose_stack import PoseStack
+
+debug = False
+
+
+class CartBondedEnergyTerm(AtomTypeDependentTerm):
+    device: torch.device  # = attr.ib()
+
+    def __init__(self, param_db: ParameterDatabase, device: torch.device):
+        super(CartBondedEnergyTerm, self).__init__(param_db=param_db, device=device)
+
+        self.cart_database = param_db.scoring.cartbonded
+        self.device = device
+
+    @classmethod
+    def score_types(cls):
+        import tmol.score.terms.cartbonded_creator
+
+        return tmol.score.terms.cartbonded_creator.CartBondedTermCreator.score_types()
+
+    def n_bodies(self):
+        return 1
+
+    def find_subgraphs(self, bonds, block_type):
+        lengths = []
+        angles = []
+        torsions = []
+        improper = []
+
+        # create a convenient datastructure for following connections
+        bondmap = {}
+        for bond in bonds:
+            if bond[0] not in bondmap.keys():
+                bondmap[bond[0]] = set()
+            bondmap[bond[0]].add(bond[1])
+
+        # get lengths
+        for atom1 in bondmap.keys():
+            for atom2 in bondmap[atom1]:
+                if atom1 < atom2:
+                    lengths.append((atom1, atom2, -1, -1))
+
+        # get angles
+        for atom1 in bondmap.keys():
+            for atom2 in bondmap[atom1]:
+                for atom3 in bondmap[atom2]:
+                    if atom1 >= atom3:
+                        continue
+                    angles.append((atom1, atom2, atom3, -1))
+
+        # get torsions
+        for atom1 in bondmap.keys():
+            for atom2 in bondmap[atom1]:
+                for atom3 in bondmap[atom2]:
+                    if atom1 == atom3:
+                        continue
+                    for atom4 in bondmap[atom3]:
+                        if atom2 == atom4:
+                            continue
+                        if atom1 >= atom4:
+                            continue
+                        torsions.append((atom1, atom2, atom3, atom4))
+
+        # get improper torsions
+        for atom3 in [block_type.atom_to_idx["CA"]]:
+            comb = list(permutations(bondmap[atom3], 3))
+            for atom1, atom2, atom4 in comb:
+                improper.append((atom1, atom2, atom3, atom4))
+
+        return (
+            lengths,
+            angles,
+            torsions,
+            improper,
+        )
```

Comment:
@aleaverfay 

This code is triggering the flake8 C901 for being too complex.

I'm not sure splitting this out into multiple functions would make it any more clear. Can I add an exception for this?